### PR TITLE
Load the bean definition instead of the bean itself

### LINKED ISSUE #6412: Micronaut startup failed: Circular dependency detected
### Expected Behavior

Micronaut application should start successfully.

### Actual Behaviour

Using Micronaut `2.5.4` Flyway migration runs into infinite loop during startup:
```
13:18:26.088 [main] INFO  i.m.flyway.AbstractFlywayMigration - Running migrations for database with qualifier [default]
13:18:26.093 [main] INFO  o.f.c.i.license.VersionPrinter - Flyway Community Edition 7.7.3 by Redgate
13:18:26.597 [main] INFO  o.f.core.internal.command.DbValidate - Successfully validated 2 migrations (execution time 00:00.017s)
13:18:26.604 [main] INFO  o.f.core.internal.command.DbMigrate - Current version of schema "portal": 1
13:18:26.604 [main] INFO  o.f.core.internal.command.DbMigrate - Schema "portal" is up to date. No migration necessary.
13:18:26.953 [main] INFO  i.m.flyway.AbstractFlywayMigration - Running migrations for database with qualifier [default]
13:18:26.954 [main] INFO  o.f.c.i.license.VersionPrinter - Flyway Community Edition 7.7.3 by Redgate
13:18:27.237 [main] INFO  o.f.core.internal.command.DbValidate - Successfully validated 2 migrations (execution time 00:00.004s)
13:18:27.239 [main] INFO  o.f.core.internal.command.DbMigrate - Current version of schema "portal": 1
13:18:27.240 [main] INFO  o.f.core.internal.command.DbMigrate - Schema "portal" is up to date. No migration necessary.
13:18:27.242 [main] INFO  i.m.flyway.AbstractFlywayMigration - Running migrations for database with qualifier [default]
13:18:27.242 [main] INFO  o.f.c.i.license.VersionPrinter - Flyway Community Edition 7.7.3 by Redgate
13:18:27.518 [main] INFO  o.f.core.internal.command.DbValidate - Successfully validated 2 migrations (execution time 00:00.004s)
13:18:27.522 [main] INFO  o.f.core.internal.command.DbMigrate - Current version of schema "portal": 1
...
...

```

Using Micronaut `3.1.1` the following error occurs during startup:
```
Message: Circular dependency detected
Path Taken: 
CompositeMeterRegistry.compositeMeterRegistry(List registries,List configurers) --> CompositeMeterRegistry.compositeMeterRegistry(List registries,[List configurers]) --> MeterRegistryConfigurer.meterRegistryConfigurer([Collection binders],Collection filters) --> MeterBinder.dataSourceMeterBinder(String dataSourceName,[DataSourcePoolMetadata dataSourcePoolMetadata]) --> TomcatDataSourcePoolMetadata.tomcatPoolDataSourceMetadataProvider([DataSource dataSource])
^                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [org.hibernate.SessionFactory] could not be loaded: Failed to inject value for parameter [configurers] of class: io.micronaut.configuration.jdbc.tomcat.metadata.TomcatDataSourcePoolMetadata
```

### Steps To Reproduce

1. Create new project using command `mn create-app example.micronaut.test --build=gradle --lang=java`
2. Add the following dependencies to the `build.gradle` file:
```
dependencies {
    annotationProcessor "io.micronaut.data:micronaut-data-processor"
    implementation("io.micronaut:micronaut-http-client")
    implementation("io.micronaut:micronaut-runtime")
    implementation("javax.annotation:javax.annotation-api")
    implementation("io.micronaut:micronaut-validation")
    implementation "io.micronaut.cache:micronaut-cache-core"
    implementation "io.micronaut.cache:micronaut-cache-caffeine"
    implementation "io.micronaut.discovery:micronaut-discovery-client"
    implementation "io.micronaut.flyway:micronaut-flyway"
    implementation "io.micronaut.data:micronaut-data-hibernate-jpa"
    implementation "io.micronaut.micrometer:micronaut-micrometer-registry-statsd"
    implementation "io.micronaut:micronaut-management"
    implementation "io.micronaut.micrometer:micronaut-micrometer-registry-prometheus"
    implementation "org.postgresql:postgresql:42.2.6"
    runtimeOnly("ch.qos.logback:logback-classic")
    runtimeOnly "io.micronaut.sql:micronaut-jdbc-tomcat"
}
```
3. Configure any simple datasource and flyway migration:
```
flyway:
  datasources:
    default:
      schemas:
        - test

datasources:
  default:
    driverClassName: org.postgresql.Driver
    dialect: POSTGRES
    url: jdbc:postgresql://localhost:5432/test
    username: test
    password: test
```    
4. Run micronaut application

### Environment Information

- Operation System: Windows 10
- Java SDK: 15

### Example Application

_No response_

### Version

2.5.4 - 3.1.1

### PR DESCRIPTION
**Pull request by jameskleeh**
Fixes #6412

### CODE REVIEW

**Review by graemerocher — APPROVED** [review]

File: runtime/src/main/java/io/micronaut/health/HeartbeatDiscoveryClientCondition.java [hunk 1]
```diff
@@ -35,8 +36,11 @@
 public final class HeartbeatDiscoveryClientCondition implements Condition {
     @Override
     public boolean matches(ConditionContext context) {
-        final CompositeDiscoveryClient compositeDiscoveryClient = context.getBean(CompositeDiscoveryClient.class);
-        final boolean hasDiscovery = compositeDiscoveryClient.getDiscoveryClients().length > 0;
+        final boolean hasDiscovery = context.getBeanContext().getBeanDefinitions(DiscoveryClient.class)
+                .stream()
+                .filter(bd -> bd.getBeanType() != (Class<?>) CompositeDiscoveryClient.class)
```
Comment (rdesgroppes) [inline runtime/src/main/java/io/micronaut/health/HeartbeatDiscoveryClientCondition.java:41]: @jameskleeh doesn't the condition always evaluate to `true`? (`CompositeDiscoveryClient` is `abstract`)

Comment (graemerocher) [reply runtime/src/main/java/io/micronaut/health/HeartbeatDiscoveryClientCondition.java:41]: Yup you are correct @rdesgroppes the check should be for `DefaultCompositeDiscoveryClient`

Comment (jameskleeh) [reply runtime/src/main/java/io/micronaut/health/HeartbeatDiscoveryClientCondition.java:41]: Good catch @rdesgroppes

Comment (rdesgroppes) [reply runtime/src/main/java/io/micronaut/health/HeartbeatDiscoveryClientCondition.java:41]: np

**Pull request merged**